### PR TITLE
[c10d] Add is_backend_available for c10d backend.

### PR DESCRIPTION
**Pull request by shaoyf42**
Add is_backend_available for c10d backend, either the built-in backends or third-party backends through function ``Backend.register_backend``.

There is a related discussion in https://github.com/pytorch/pytorch/pull/101775#discussion_r1199253553
> For example in python constructor for their backend they should explicitly add the is_X_available. Or if defining in C++ they should modify pybind like this https://github.com/H-Huang/torch_collective_extension/blob/main/custom_backend/include/dummy.hpp#L98-L101
to also add their own is_available property

It is a natural choice for users to add their own `is_available` when they create a backend. We think it might be a possible way for the user to use `is_X_available` in the same way as the native, for example by dynamically adding`torch.distributed.is_dummpy_available()` function.  This is why we want to dynamically add the `is_X_available` to `torch.distributed` in `register_backend`.

> Or we could add an Is_available(backend) function, that checks for the backend.

Providing a public function is indeed another good approach. We have implemented an `is_backend_available` in https://github.com/pytorch/pytorch/pull/101945  that supports both built-in backends and third-party backends.

### CODE REVIEW

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -896,6 +896,24 @@ def is_ucc_available() -> bool:
     return _UCC_AVAILABLE
 
 
+def is_backend_available(backend: str) -> bool:
```
Comment (shaoyf42) [inline torch/distributed/distributed_c10d.py:899]: Here,  `is_backend_available` is added specifically for checking whether the backend is available or not. Adding a backend parameter to `torch.distributed.is_available` might be another possible way to do this.

**Comment by shaoyf42** [comment]
@H-Huang Providing an `is_backend_available` seems to be a better way than the user manually adding the `is_dummy_available` to `torch.distributed`. Could you take a look?

**Review by H-Huang — APPROVED** [review]
Thanks for adding

**Pull request closed without merging**